File: 2419_LongestSubarrayWithMaximumBitwiseAND.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        ios_base::sync_with_stdio(0);cin.tie(0);cout.tie(0);
        
        int maxNumber=0, maxLength = 1, n= nums.size(), i=0, runningLength, curr;

        while (i<n){
            curr = nums[i++];
            runningLength=1;
            while (i<n && nums[i]==curr){
                ++runningLength;
                ++i;
            }
            if (maxNumber<curr){
                maxLength = runningLength;
                maxNumber= curr;
            } else if (maxNumber==curr){
                maxLength = max(maxLength, runningLength);
            }
        }
        return maxLength;
    }
};

int main(){
    Solution s;
    vector<int> nums = {100,5,5};
    cout<<s.longestSubarray(nums);
    return 0;
}

File: 2419_LongestSubarrayWithMaximumBitwiseAND.py
from typing import List


#O(n)
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 1

        maxNum, currLen = nums[0], 1

        for i in range(1,n):
            if nums[i]>maxNum:
                ans = currLen = 1
                maxNum = nums[i]
            elif nums[i]==maxNum:
                if nums[i]==nums[i-1]:
                    currLen+=1
                    ans = max(ans, currLen)
                else:
                    currLen=1

        return ans
    

# #O(n)
# class Solution:
#     def longestSubarray(self, nums: List[int]) -> int:
#         n = len(nums)
#         ans = 1

#         currNum, maxNum, left = nums[0], nums[0], 0
#         for right in range(1,n):
#             if nums[right] != nums[left]:
#                 currNum = nums[right]
#                 left = right

#             if currNum==maxNum:
#                 ans = max(ans, right-left+1)
#             elif currNum>maxNum:
#                 maxNum = currNum
#                 ans = right-left+1

#         return ans
